libnetwork/drivers/overlay/ovmanager/ovmanager.go

```diff
@@ -111,11 +109,18 @@ func (d *driver) NetworkAllocate(id string, option map[string]string, ipV4Data,
 
 		if len(vxlanIDList) > i {
 			s.vni = vxlanIDList[i]
-		}
-
-		if err := n.obtainVxlanID(s); err != nil {
-			n.releaseVxlanID()
-			return nil, fmt.Errorf("could not obtain vxlan id for pool %s: %v", s.subnetIP, err)
+			err := d.vxlanIdm.Set(uint64(s.vni))
+			if err != nil {
+				n.releaseVxlanID()
+				return nil, fmt.Errorf("could not assign vxlan id %v to pool %s: %v", s.vni, s.subnetIP, err)
+			}
+		} else {
```

Comment:
It was a little tricky to figure out what the existing logic was doing, but for anyone else reviewing:

`s.vni` would have been the zero value due to the earlier `len(vxlanIDList) > i` conditional by the time we hit the old `obtainVxlanID`, so this is a transparent simplification.

Reply:
Maybe lets get a comment here?
This is pretty difficult to understand at a glance.

Reply:
Also to note, I think the logic above this loop (which does the `strings.Split`) is fairly expensive (or at least has potential to be).

Reply:
I honestly don't understand why `vxlanIDList` corresponds to anything in `ipv4Data`, it would be great to have a comment around this... or for that matter why the5re isn't something similar for ipv6 (other than I guess its just not implemented in overlay?)

Reply:
> Also to note, I think the logic above this loop (which does the `strings.Split`) is fairly expensive (or at least has potential to be).

The lock is not taken until after that logic.

> I honestly don't understand why `vxlanIDList` corresponds to anything in `ipv4Data`, it would be great to have a comment around this...

I don't understand why a separate VNI is allocated per subnet either. And in true libnetwork fashion, the original rationale for _why_ it was [implemented that way](https://github.com/moby/libnetwork/pull/584) has been lost to the mists of time. All I can do is speculate.

> or for that matter why the5re isn't something similar for ipv6 (other than I guess its just not implemented in overlay?)

- It's not implemented: #43615

Reply:
> The lock is not taken until after that logic.

I'm more referring to the potentially unbounded slice that gets created over using an approach that scans over the string.
It may even be possible to use repeated `strings.Cut` calls in the for loop.

Reply:
Fair enough. That's certainly worth addressing, though it's out of scope of this PR. The similar code in the non-manager overlay driver should also be updated at the same time. https://github.com/moby/moby/blob/11c6ec6eb2033600528edcb08c57a88d45dc2e9a/libnetwork/drivers/overlay/ov_network.go#L108-L122

Reply:
https://github.com/moby/moby/pull/45892